File: tui/src/scala/tui/widgets/TabsWidget.scala
package tui
package widgets
package tabs

import tui.internal.ranges
import tui.internal.saturating._

/** A widget to display available tabs in a multiple panels context.
  *
  * @param block
  *   A block to wrap this widget in if necessary
  * @param titles
  *   One title for each tab
  * @param selected
  *   The index of the selected tabs
  * @param style
  *   The style used to draw the text
  * @param highlightStyle
  *   Style to apply to the selected item
  * @param divider
  *   Tab divider
  */
case class TabsWidget(
    titles: Array[Spans],
    selected: Int = 0,
    style: Style = Style.DEFAULT,
    highlightStyle: Style = Style.DEFAULT,
    divider: Span = Span.nostyle(symbols.line.VERTICAL)
) extends Widget {

  def render(area: Rect, buf: Buffer): Unit = {
    buf.setStyle(area, style)
    if (area.height < 1) {
      return
    }

    var x = area.left
    val titles_length = titles.length
    ranges.range(0, titles_length) { i =>
      val title = titles(i)
      val last_title = titles_length - 1 == i
      x = x.saturating_add(1)
      val remaining_width = area.right.saturating_sub_unsigned(x)
      if (remaining_width == 0) {
        ()
      } else {
        val pos = buf.setSpans(x, area.top, title, remaining_width)
        if (i == selected) {
          buf.setStyle(
            Rect(
              x,
              y = area.top,
              width = pos._1.saturating_sub_unsigned(x),
              height = 1
            ),
            highlightStyle
          )
        }
        x = pos._1.saturating_add(1)
        val remaining_width1 = area.right.saturating_sub_unsigned(x)
        if (remaining_width1 == 0 || last_title) {
          ()
        } else {
          val pos = buf.setSpan(x, area.top, divider, remaining_width1)
          x = pos._1
        }
      }
    }
  }
}

File: tui/src/java/tui/widgets/tabs/TabsWidget.java
package tui.widgets.tabs;

import java.util.Optional;
import tui.Buffer;
import tui.Position;
import tui.Rect;
import tui.Span;
import tui.Spans;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;
import tui.internal.Saturating;
import tui.widgets.BlockWidget;

/// A widget to display available tabs in a multiple panels context.
public final class TabsWidget implements Widget {
  public final Optional<BlockWidget> block;
  public final Spans[] titles;
  public final int selected;
  public final Style style;
  public final Style highlightStyle;
  public final Span divider;

  public TabsWidget(
      Optional<BlockWidget> block,
      Spans[] titles,
      int selected,
      Style style,
      Style highlightStyle,
      Span divider) {
    this.block = block;
    this.titles = titles;
    this.selected = selected;
    this.style = style;
    this.highlightStyle = highlightStyle;
    this.divider = divider;
  }

  public static TabsWidget empty(Spans[] titles) {
    return new TabsWidget(
        Optional.empty(),
        titles,
        0,
        Style.DEFAULT,
        Style.DEFAULT,
        Span.nostyle(Symbols.line.VERTICAL));
  }

  public TabsWidget withBlock(BlockWidget b) {
    return new TabsWidget(Optional.of(b), titles, selected, style, highlightStyle, divider);
  }

  public TabsWidget withTitles(Spans[] t) {
    return new TabsWidget(block, t, selected, style, highlightStyle, divider);
  }

  public TabsWidget withSelected(int s) {
    return new TabsWidget(block, titles, s, style, highlightStyle, divider);
  }

  public TabsWidget withStyle(Style s) {
    return new TabsWidget(block, titles, selected, s, highlightStyle, divider);
  }

  public TabsWidget withHighlightStyle(Style s) {
    return new TabsWidget(block, titles, selected, style, s, divider);
  }

  public TabsWidget withDivider(Span d) {
    return new TabsWidget(block, titles, selected, style, highlightStyle, d);
  }

  @Override
  public void render(Rect area, Buffer buf) {
    buf.setStyle(area, style);
    Rect tabsArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      tabsArea = innerArea;
    } else {
      tabsArea = area;
    }

    if (tabsArea.height() < 1) {
      return;
    }

    int[] x = {tabsArea.left()};
    int titlesLength = titles.length;
    Ranges.range(
        0,
        titlesLength,
        i -> {
          Spans title = titles[i];
          boolean lastTitle = titlesLength - 1 == i;
          x[0] = Saturating.saturatingAdd(x[0], 1);
          int remainingWidth = Saturating.saturatingSubUnsigned(tabsArea.right(), x[0]);
          if (remainingWidth == 0) {
            return;
          }
          Position pos = buf.setSpans(x[0], tabsArea.top(), title, remainingWidth);
          if (i == selected) {
            buf.setStyle(
                new Rect(
                    x[0],
                    tabsArea.top(),
                    Saturating.saturatingSubUnsigned(pos.x(), x[0]),
                    1),
                highlightStyle);
          }
          x[0] = Saturating.saturatingAdd(pos.x(), 1);
          int remainingWidth1 = Saturating.saturatingSubUnsigned(tabsArea.right(), x[0]);
          if (remainingWidth1 == 0 || lastTitle) {
            return;
          }
          Position pos2 = buf.setSpan(x[0], tabsArea.top(), divider, remainingWidth1);
          x[0] = pos2.x();
        });
  }
}
